File: polydiff/parsers/jdk_uri/Wrapper.java
// PolyDiff parser wrapper for java.net.URI (JDK 21).
//
// Subprocess oracle, per SPEC §5.3.
//   stdin  : up to 64 KiB UTF-8
//   stdout : v2 fact-vector envelope (one JSON line)
//   exit 0 : parse-or-error (parser's verdict)
//   exit !=0 : wrapper crash
//
// This wrapper uses ONLY the JDK standard library (no Maven, no Gradle,
// no third-party jars) so it builds in the pinned devcontainer with
// just `javac`.
//
// Compile:
//   javac Wrapper.java
// Run:
//   echo -n 'https://example.com/foo' | java -cp . Wrapper --input-id ID

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.List;

public class Wrapper {
    private static final int MAX_BYTES = 64 * 1024;

    public static void main(String[] args) throws IOException {
        String inputId = parseInputId(args);
        String raw = readStdin();
        String json = parse(inputId, raw);
        System.out.println(json);
    }

    private static String parseInputId(String[] args) {
        for (int i = 0; i + 1 < args.length; i++) {
            if ("--input-id".equals(args[i])) {
                return args[i + 1];
            }
        }
        throw new IllegalArgumentException("missing --input-id");
    }

    private static String readStdin() throws IOException {
        ByteArrayOutputStream buf = new ByteArrayOutputStream();
        InputStream in = System.in;
        byte[] chunk = new byte[4096];
        int total = 0;
        int n;
        while ((n = in.read(chunk)) != -1 && total < MAX_BYTES) {
            int allow = Math.min(n, MAX_BYTES - total);
            buf.write(chunk, 0, allow);
            total += allow;
        }
        return buf.toString(java.nio.charset.StandardCharsets.UTF_8);
    }

    private static String parse(String inputId, String raw) {
        List<String> errors = new ArrayList<>();
        List<String> warnings = new ArrayList<>();
        warnings.add("axis 'host_has_idn' not directly observable by parser 'jdk_uri'");
        warnings.add("axis 'host_punycode' not directly observable by parser 'jdk_uri'");
        warnings.add("axis 'control_chars_in_host_rejected' not directly observable by parser 'jdk_uri'");

        URI uri;
        try {
            uri = new URI(raw);
        } catch (URISyntaxException ex) {
            errors.add("java.net.URI: " + ex.getMessage());
            return emptyEnvelope(inputId, errors, warnings);
        }

        String scheme = uri.getScheme();
        String host = uri.getHost();
        // java.net.URI does NOT lowercase host on construction; getHost()
        // returns case-preserving.
        String userInfo = uri.getUserInfo();
        String username = null;
        boolean passwordPresent = false;
        if (userInfo != null) {
            int colon = userInfo.indexOf(':');
            if (colon >= 0) {
                username = userInfo.substring(0, colon);
                passwordPresent = colon + 1 < userInfo.length();
            } else {
                username = userInfo;
            }
        }
        int rawPort = uri.getPort();
        Integer port = (rawPort == -1) ? null : rawPort;
        String path = uri.getRawPath();
        String query = uri.getRawQuery();
        String fragment = uri.getRawFragment();

        boolean userinfoPresent = userInfo != null;
        boolean isIPv4 = host != null && isIPv4Literal(host);
        boolean isIPv6 = host != null && (host.startsWith("[") && host.endsWith("]"));
        boolean isIpLit = isIPv4 || isIPv6;
        boolean isLoopback = isIPv4 && host.startsWith("127.");
        boolean isPrivate = isPrivateOrLinkLocal(host);

        StringBuilder b = new StringBuilder(2048);
        b.append('{');
        appendKV(b, "schema_version", "v2"); b.append(',');
        appendKV(b, "input_id", inputId); b.append(',');
        appendKV(b, "parser_profile", "jdk_uri"); b.append(',');
        appendBool(b, "parsed", true); b.append(',');
        appendStringArray(b, "errors", errors); b.append(',');
        appendStringArray(b, "warnings", warnings); b.append(',');
        appendKV(b, "scheme", scheme); b.append(',');
        appendKV(b, "scheme_lowercased", scheme == null ? null : scheme.toLowerCase()); b.append(',');
        appendBool(b, "userinfo_present", userinfoPresent); b.append(',');
        appendKV(b, "userinfo_raw", userInfo); b.append(',');
        appendKV(b, "username", username); b.append(',');
        appendBool(b, "password_present", passwordPresent); b.append(',');
        appendKV(b, "host", host); b.append(',');
        appendKV(b, "host_raw", host); b.append(',');
        appendKV(b, "host_lowercased", host == null ? null : host.toLowerCase()); b.append(',');
        appendKV(b, "host_decoded", host); b.append(',');
        appendBool(b, "host_is_ip_literal", host == null ? null : isIpLit); b.append(',');
        appendBool(b, "host_is_ipv4", host == null ? null : isIPv4); b.append(',');
        appendBool(b, "host_is_ipv6", host == null ? null : isIPv6); b.append(',');
        appendBool(b, "host_is_ipvFuture", host == null ? null : false); b.append(',');
        appendBool(b, "host_is_loopback", host == null ? null : isLoopback); b.append(',');
        appendBool(b, "host_is_private_or_link_local", isPrivate); b.append(',');
        appendNull(b, "host_has_idn"); b.append(',');
        appendNull(b, "host_punycode"); b.append(',');
        appendInt(b, "port", port); b.append(',');
        appendBool(b, "port_present", port != null); b.append(',');
        appendInt(b, "port_value", port); b.append(',');
        appendInt(b, "port_default_inferred", defaultPortFor(scheme)); b.append(',');
        appendKV(b, "path", path); b.append(',');
        appendKV(b, "path_raw", path); b.append(',');
        appendKV(b, "path_normalized", path); b.append(',');
        appendBool(b, "path_traversal_resolved", false); b.append(',');
        appendKV(b, "query_raw", query); b.append(',');
        appendKey(b, "query_pairs"); b.append("[]"); b.append(',');
        appendKV(b, "fragment_raw", fragment); b.append(',');
        appendBool(b, "percent_decoding_applied_in_host", false); b.append(',');
        appendBool(b, "percent_decoding_applied_in_path", false); b.append(',');
        appendBool(b, "trailing_slash_normalized", false); b.append(',');
        appendBool(b, "leading_zeroes_in_octets_stripped", false); b.append(',');
        // java.net.URI does NOT strip tabs/newlines; it raises URISyntaxException
        // on most control chars, but not all; report null to be honest.
        appendNull(b, "tab_or_newline_stripped"); b.append(',');
        appendBool(b, "backslash_treated_as_slash", false); b.append(',');
        appendNull(b, "control_chars_in_host_rejected"); b.append(',');
        appendBool(b, "scheme_authority_separator_strict", true); b.append(',');
        appendKV(b, "raw_serialized", uri.toString()); b.append(',');
        appendKV(b, "parse_error", null);
        b.append('}');
        return b.toString();
    }

    private static String emptyEnvelope(String inputId, List<String> errors, List<String> warnings) {
        StringBuilder b = new StringBuilder(1024);
        b.append('{');
        appendKV(b, "schema_version", "v2"); b.append(',');
        appendKV(b, "input_id", inputId); b.append(',');
        appendKV(b, "parser_profile", "jdk_uri"); b.append(',');
        appendBool(b, "parsed", false); b.append(',');
        appendStringArray(b, "errors", errors); b.append(',');
        appendStringArray(b, "warnings", warnings); b.append(',');
        for (String key : new String[]{
                "scheme","scheme_lowercased","userinfo_raw","username","host","host_raw",
                "host_lowercased","host_decoded","host_punycode","path","path_raw",
                "path_normalized","query_raw","fragment_raw","raw_serialized",
                "parse_error"
        }) {
            appendKV(b, key, key.equals("parse_error") ? (errors.isEmpty() ? null : errors.get(0)) : null);
            b.append(',');
        }
        for (String key : new String[]{
                "userinfo_present","password_present","host_is_private_or_link_local",
                "trailing_slash_normalized","percent_decoding_applied_in_host",
                "percent_decoding_applied_in_path","leading_zeroes_in_octets_stripped",
                "backslash_treated_as_slash"
        }) {
            appendBool(b, key, false); b.append(',');
        }
        for (String key : new String[]{
                "host_is_ip_literal","host_is_ipv4","host_is_ipv6","host_is_ipvFuture",
                "host_is_loopback","host_has_idn","port_present",
                "path_traversal_resolved","tab_or_newline_stripped",
                "control_chars_in_host_rejected","scheme_authority_separator_strict"
        }) {
            appendNull(b, key); b.append(',');
        }
        for (String key : new String[]{"port","port_value","port_default_inferred"}) {
            appendInt(b, key, null); b.append(',');
        }
        appendKey(b, "query_pairs"); b.append("[]");
        b.append('}');
        return b.toString();
    }

    private static boolean isIPv4Literal(String host) {
        if (host == null || host.isEmpty()) return false;
        String[] parts = host.split("\\.");
        if (parts.length != 4) return false;
        for (String p : parts) {
            if (p.isEmpty() || p.length() > 3) return false;
            for (int i = 0; i < p.length(); i++) {
                if (!Character.isDigit(p.charAt(i))) return false;
            }
            int v;
            try { v = Integer.parseInt(p); } catch (NumberFormatException e) { return false; }
            if (v < 0 || v > 255) return false;
        }
        return true;
    }

    private static boolean isPrivateOrLinkLocal(String host) {
        if (host == null) return false;
        if (host.startsWith("[")) return false;
        if (!isIPv4Literal(host)) return false;
        String[] parts = host.split("\\.");
        int o0 = Integer.parseInt(parts[0]);
        int o1 = Integer.parseInt(parts[1]);
        if (o0 == 10) return true;
        if (o0 == 127) return true;
        if (o0 == 172 && o1 >= 16 && o1 <= 31) return true;
        if (o0 == 192 && o1 == 168) return true;
        if (o0 == 169 && o1 == 254) return true;
        return false;
    }

    private static Integer defaultPortFor(String scheme) {
        if (scheme == null) return null;
        switch (scheme.toLowerCase()) {
            case "http": return 80;
            case "https": return 443;
            case "ws": return 80;
            case "wss": return 443;
            case "ftp": return 21;
            default: return null;
        }
    }

    private static void appendKey(StringBuilder b, String k) {
        b.append('"').append(escape(k)).append('"').append(':');
    }
    private static void appendKV(StringBuilder b, String k, String v) {
        appendKey(b, k);
        if (v == null) b.append("null");
        else b.append('"').append(escape(v)).append('"');
    }
    private static void appendBool(StringBuilder b, String k, Boolean v) {
        appendKey(b, k);
        if (v == null) b.append("null");
        else b.append(v.booleanValue() ? "true" : "false");
    }
    private static void appendInt(StringBuilder b, String k, Integer v) {
        appendKey(b, k);
        if (v == null) b.append("null");
        else b.append(v.intValue());
    }
    private static void appendNull(StringBuilder b, String k) {
        appendKey(b, k); b.append("null");
    }
    private static void appendStringArray(StringBuilder b, String k, List<String> v) {
        appendKey(b, k); b.append('[');
        for (int i = 0; i < v.size(); i++) {
            if (i > 0) b.append(',');
            b.append('"').append(escape(v.get(i))).append('"');
        }
        b.append(']');
    }

    private static String escape(String s) {
        if (s == null) return "";
        StringBuilder out = new StringBuilder(s.length() + 8);
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"': out.append("\\\""); break;
                case '\\': out.append("\\\\"); break;
                case '\n': out.append("\\n"); break;
                case '\r': out.append("\\r"); break;
                case '\t': out.append("\\t"); break;
                default:
                    if (c < 0x20) {
                        out.append(String.format("\\u%04x", (int) c));
                    } else if (c > 0x7E) {
                        out.append(String.format("\\u%04x", (int) c));
                    } else {
                        out.append(c);
                    }
            }
        }
        return out.toString();
    }
}

File: polydiff/parsers/python_urllib/wrapper.py
#!/usr/bin/env python3
"""PolyDiff parser wrapper for CPython 3.11+ stdlib urllib.parse.

Subprocess oracle, per SPEC §5.3:

  stdin  : up to 64 KiB of raw bytes (the candidate URL string, UTF-8)
  stdout : exactly one JSON line; v2 fact-vector envelope
  exit 0 : parse-or-error (parser's verdict, not infrastructure failure)
  exit !0: wrapper crash (handled as a Finding by the orchestrator)

This wrapper deliberately does NOT import any 577 Industries module. It
uses only the Python standard library so it can be containerized (or run
with a different CPython interpreter) without dragging the rest of the
repo along.

Read input_id from CLI arg --input-id (the orchestrator passes it
in). The URL itself comes from stdin to keep the contract uniform
across language runtimes.
"""

from __future__ import annotations

import argparse
import ipaddress
import json
import sys
import urllib.parse


def _safe_port(parsed: urllib.parse.SplitResult) -> int | None:
    try:
        return parsed.port
    except ValueError:
        return None


def _is_ipv4(host: str | None) -> bool | None:
    if host is None:
        return None
    try:
        ipaddress.IPv4Address(host)
        return True
    except (ipaddress.AddressValueError, ValueError):
        return False


def _is_ipv6(host: str | None) -> bool | None:
    if host is None:
        return None
    candidate = host.strip("[]")
    try:
        ipaddress.IPv6Address(candidate)
        return True
    except (ipaddress.AddressValueError, ValueError):
        return False


def _is_private_or_link_local(host: str | None) -> bool:
    if not host:
        return False
    candidate = host.strip("[]")
    try:
        ip = ipaddress.ip_address(candidate)
    except ValueError:
        return False
    return ip.is_private or ip.is_loopback or ip.is_link_local


def _is_loopback(host: str | None) -> bool | None:
    if host is None:
        return None
    candidate = host.strip("[]")
    try:
        ip = ipaddress.ip_address(candidate)
    except ValueError:
        # Hostname-based loopback (e.g. "localhost") is not directly observable
        # from urllib.parse without resolution; report null to avoid lying.
        if candidate.lower() == "localhost":
            return True
        return False
    return ip.is_loopback


def parse(input_id: str, raw_url: str) -> dict[str, object]:
    """Parse `raw_url` with urllib.parse and emit a v2 fact vector."""
    errors: list[str] = []
    warnings: list[str] = [
        "axis 'host_has_idn' not directly observable by parser 'python_urllib'",
        "axis 'host_punycode' not directly observable by parser 'python_urllib'",
        "axis 'control_chars_in_host_rejected' not directly observable by parser 'python_urllib'",
    ]
    parsed_ok = True

    try:
        parsed = urllib.parse.urlsplit(raw_url)
    except ValueError as exc:
        errors.append(f"urllib.parse.urlsplit raised ValueError: {exc}")
        parsed_ok = False
        return _empty_envelope(input_id, raw_url, errors, warnings)

    scheme = parsed.scheme or None
    host = parsed.hostname  # already lowercased by urllib
    port = _safe_port(parsed)
    if port is None and parsed.netloc:
        # urllib raises on out-of-range ports; treat as wrapper-visible error.
        if ":" in parsed.netloc.split("@", 1)[-1]:
            errors.append("urllib could not parse port (out-of-range or non-numeric)")

    userinfo_raw = None
    if parsed.username is not None or parsed.password is not None:
        # Reconstruct what was between scheme and host
        netloc = parsed.netloc
        if "@" in netloc:
            userinfo_raw = netloc.rsplit("@", 1)[0]

    is_ip4 = _is_ipv4(host)
    is_ip6 = _is_ipv6(host) or _is_ipv6(parsed.hostname or "")

    # Compute input-driven axes ("did this input exhibit X feature, and did
    # the parser act on it?"). This is materially different from
    # spec-conformance flags — we only report a boolean if the relevant
    # feature was present in the input. Otherwise: None ("no opinion").
    input_has_tab_or_newline = any(c in raw_url for c in "\t\n\r")
    input_has_backslash = "\\" in raw_url
    input_has_dotdot = "/../" in raw_url or raw_url.endswith("/..") or "/." in raw_url
    serialized = parsed.geturl()

    tab_or_newline_stripped: bool | None = None
    if input_has_tab_or_newline:
        tab_or_newline_stripped = not any(c in serialized for c in "\t\n\r")

    backslash_treated_as_slash: bool | None = None
    if input_has_backslash:
        backslash_treated_as_slash = "\\" not in serialized

    path_traversal_resolved: bool | None = None
    if input_has_dotdot:
        path_traversal_resolved = "/.." not in (parsed.path or "") and "/." not in (parsed.path or "")

    leading_zeroes_in_octets_stripped: bool | None = None
    # Heuristic: input host had a leading-zero IPv4 octet (e.g. 0177.0.0.1)
    # AND the parser produced a host without it.
    raw_host_in_input = ""
    if "://" in raw_url:
        after_scheme = raw_url.split("://", 1)[1]
        raw_host_in_input = after_scheme.split("/", 1)[0].split("?", 1)[0].split("#", 1)[0].split("@")[-1].split(":")[0]
    if raw_host_in_input and any(p.startswith("0") and len(p) > 1 and p.isdigit() for p in raw_host_in_input.split(".")):
        leading_zeroes_in_octets_stripped = host is not None and not any(
            p.startswith("0") and len(p) > 1 and p.isdigit() for p in (host or "").split(".")
        )

    percent_decoding_applied_in_host: bool | None = None
    if "%" in raw_host_in_input:
        percent_decoding_applied_in_host = host is not None and "%" not in host

    percent_decoding_applied_in_path: bool | None = None
    if parsed.path and "%" in parsed.path:
        # urllib does not percent-decode path on parse; stays raw
        percent_decoding_applied_in_path = False

    control_chars_in_host_rejected: bool | None = None
    if any(0 <= ord(c) < 0x20 for c in (raw_host_in_input or "")):
        # urllib silently keeps control chars in host
        control_chars_in_host_rejected = False

    return {
        "schema_version": "v2",
        "input_id": input_id,
        "parser_profile": "python_urllib",
        "parsed": parsed_ok,
        "errors": errors,
        "warnings": warnings,
        "scheme": scheme,
        "scheme_lowercased": scheme.lower() if scheme else None,
        "userinfo_present": bool(parsed.username or parsed.password),
        "userinfo_raw": userinfo_raw,
        "username": parsed.username,
        "password_present": parsed.password is not None,
        "host": host,
        "host_raw": parsed.hostname,
        "host_lowercased": host.lower() if host else None,
        "host_decoded": host,  # urllib does no IDN decoding
        "host_is_ip_literal": (is_ip4 or is_ip6) if host else None,
        "host_is_ipv4": is_ip4,
        "host_is_ipv6": is_ip6,
        "host_is_ipvFuture": False,
        "host_is_loopback": _is_loopback(host),
        "host_is_private_or_link_local": _is_private_or_link_local(host),
        "host_has_idn": None,
        "host_punycode": None,
        "port": port,
        "port_present": port is not None,
        "port_value": port,
        "port_default_inferred": _default_port_for(scheme),
        "path": parsed.path or None,
        "path_raw": parsed.path or None,
        "path_normalized": _normalize_path(parsed.path),
        "path_traversal_resolved": path_traversal_resolved,
        "query_raw": parsed.query or None,
        "query_pairs": _parse_query(parsed.query),
        "fragment_raw": parsed.fragment or None,
        "percent_decoding_applied_in_host": percent_decoding_applied_in_host,
        "percent_decoding_applied_in_path": percent_decoding_applied_in_path,
        "trailing_slash_normalized": None,
        "leading_zeroes_in_octets_stripped": leading_zeroes_in_octets_stripped,
        "tab_or_newline_stripped": tab_or_newline_stripped,
        "backslash_treated_as_slash": backslash_treated_as_slash,
        "control_chars_in_host_rejected": control_chars_in_host_rejected,
        "scheme_authority_separator_strict": True,
        "raw_serialized": serialized,
        "parse_error": None if parsed_ok else (errors[0] if errors else "parse_error"),
    }


def _empty_envelope(
    input_id: str, raw_url: str, errors: list[str], warnings: list[str]
) -> dict[str, object]:
    return {
        "schema_version": "v2",
        "input_id": input_id,
        "parser_profile": "python_urllib",
        "parsed": False,
        "errors": errors,
        "warnings": warnings,
        "scheme": None,
        "scheme_lowercased": None,
        "userinfo_present": False,
        "userinfo_raw": None,
        "username": None,
        "password_present": False,
        "host": None,
        "host_raw": None,
        "host_lowercased": None,
        "host_decoded": None,
        "host_is_ip_literal": None,
        "host_is_ipv4": None,
        "host_is_ipv6": None,
        "host_is_ipvFuture": None,
        "host_is_loopback": None,
        "host_is_private_or_link_local": False,
        "host_has_idn": None,
        "host_punycode": None,
        "port": None,
        "port_present": None,
        "port_value": None,
        "port_default_inferred": None,
        "path": None,
        "path_raw": None,
        "path_normalized": None,
        "path_traversal_resolved": None,
        "query_raw": None,
        "query_pairs": None,
        "fragment_raw": None,
        "percent_decoding_applied_in_host": None,
        "percent_decoding_applied_in_path": None,
        "trailing_slash_normalized": None,
        "leading_zeroes_in_octets_stripped": None,
        "tab_or_newline_stripped": None,
        "backslash_treated_as_slash": None,
        "control_chars_in_host_rejected": None,
        "scheme_authority_separator_strict": None,
        "raw_serialized": None,
        "parse_error": (errors[0] if errors else "parse_error"),
    }


def _default_port_for(scheme: str | None) -> int | None:
    if not scheme:
        return None
    return {"http": 80, "https": 443, "ws": 80, "wss": 443, "ftp": 21}.get(scheme.lower())


def _normalize_path(raw: str | None) -> str | None:
    if raw is None or raw == "":
        return raw
    # urllib.parse does not resolve `.` / `..`; just return what it gave us.
    return raw


def _parse_query(query: str) -> list[dict[str, str]] | None:
    if not query:
        return [] if query == "" else None
    pairs: list[dict[str, str]] = []
    for raw in query.split("&"):
        if "=" in raw:
            k, v = raw.split("=", 1)
        else:
            k, v = raw, ""
        pairs.append({"key": k, "value": v})
    return pairs


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--input-id", required=True)
    args = parser.parse_args(argv)

    raw_bytes = sys.stdin.buffer.read(64 * 1024)
    try:
        raw_url = raw_bytes.decode("utf-8", errors="replace")
    except Exception as exc:  # pragma: no cover - decode("replace") never raises
        print(json.dumps({"error": f"decode failed: {exc}"}))
        return 2

    fv = parse(args.input_id, raw_url)
    sys.stdout.write(json.dumps(fv, ensure_ascii=True))
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())
